### ISSUE TITLE
Wiki Documentation Update

### ISSUE THREAD

**Issue by CraigSelbert** [dotnet/roslyn#8803]
On the wiki page - [Getting Started C# Semantic Analysis](https://github.com/dotnet/roslyn/wiki/Getting-Started-C%23-Semantic-Analysis), under "Understanding Compilations and Symbols" - step 3 it currently is using "MetadataReference.CreateFromAssembly" which has been deprecated. Please change 

`var compilation = CSharpCompilation.Create("HelloWorld")
                                               .AddReferences(
                                                    MetadataReference.CreateFromAssembly(
                                                        typeof(object).Assembly))
                                               .AddSyntaxTrees(tree);
 `

to

`var compilation = CSharpCompilation.Create("HelloWorld")
                                               .AddReferences(
                                                    MetadataReference.CreateFromFile(
                                                        typeof(object).Assembly.Location))
                                               .AddSyntaxTrees(tree);
 `

**Comment by npolyak** [comment]
Why was the functionality deprecated? For example I have an in-memory dynamically compiled and loaded assembly. How do I add it to the compilation?

**Comment by tmat** [comment]
I have fixed the wiki page to use ```CreateFromFile```. Keeping this issue around for tracking "un-deprecation" of the original API. The API will be needed in order to remove dependency on compiler internals from scripting.

**Comment by npolyak** [comment]
I am afraid that 'un-deprecation' is not sufficient - creation of reference from dynamic assembly will have to be implemented - here is the code of MetadataReference.CreateFromAssemblyInternal that throws an exception if assembly is dynamic or its location is null or empty:

`            if (assembly == null)
            {
                throw new ArgumentNullException(nameof(assembly));
            }

            if (assembly.IsDynamic)
            {
                throw new NotSupportedException(CodeAnalysisResources.CantCreateReferenceToDynamicAssembly);
            }

            if (properties.Kind != MetadataImageKind.Assembly)
            {
                throw new ArgumentException(CodeAnalysisResources.CantCreateModuleReferenceToAssembly, nameof(properties));
            }

            string location = CorLightup.Desktop.GetAssemblyLocation(assembly);
            if (string.IsNullOrEmpty(location))
            {
                throw new NotSupportedException(CodeAnalysisResources.CantCreateReferenceToAssemblyWithoutLocation);
            }`

**Comment by tmat** [comment]
@npolyak Yes, you're right. Additional work is needed. Not all dynamic assemblies will be supported on all platforms though any time soon though. 

There is a difference between "in-memory dynamically compiled assembly" and "dynamic assembly" (IsDynamic returning true). The former is an assembly compiled at runtime (dynamically) into a byte[]. The latter is an AssemblyBuilder assembly created by Reflection.Emit. 

The CLR implemented new API to access metadata of an assembly object (https://github.com/dotnet/corefx/issues/2768). This API is currently only available on Core CLR (afaik). We could easily use that API to load the byte[] assembly. Still we won't be able to get metadata for the AssemblyBuilder. Compiling against AssemblyBuilder would require implementing a whole new set of Roslyn PE symbols that could wrap Reflection objects, which would be very expensive.

**Comment by npolyak** [comment]
Hey Tomas 
even doing one of them would be a great step forward.
Thanks
Nick

**Comment by tmat** [comment]
Orleans would benefit from this API: https://github.com/dotnet/orleans/blob/22c2a63ab68e94417541a09e074ab6aeb697303f/src/OrleansCodeGenerator/CodeGeneratorCommon.cs

**Comment by redwyre** [comment]
I'm using Costura so I had to re-decompress the assembly through a compression stream, to a memory stream, to a byte array to be able to load it.

If I'm going to be loading the generated assembly it's going to reference the in-memory assemblies, so it would make sense that I should be able to reference them directly in roslyn...
